src/pcre2_compile.c

```diff
@@ -1906,15 +1907,15 @@ else
       \1 to \9 are always back references. \8x and \9x are too; \1x to \7x
       are octal escapes if there are not that many previous captures. */
 
-      if (read_number(&ptr, ptrend, -1, INT_MAX/10 - 1, 0, &s, errorcodeptr) &&
+      if (read_number(&ptr, ptrend, -1, INT_MAX/10 - 1, ERR61, &s, errorcodeptr) &&
           (s < 10 || oldptr[-1] >= CHAR_8 || s <= (int)cb->bracount))
         {
         if (s > (int)MAX_GROUP_NUMBER) *errorcodeptr = ERR61;
           else escape = -s;     /* Indicates a back reference */
         break;
         }
 
-      ptr = oldptr;      /* Put the pointer back and fall through */
+      if (c <= CHAR_7) ptr = oldptr; /* Put the pointer back for fall through */
```

Comment:
Okay... this seems convoluted to me. The check for `oldptr[-1] >= CHAR_8` was previously done in one single place. Now you've effectively duplicated that check in two locations (and inverted the logic here).

And, you've changed it so that `errorcodeptr` state is now polluted, but is later conditionally un-polluted, with a third check on `if (c >= CHAR_8) break`.

Personally, that seems less clear.

But, I have tested the code with all the possible scenarios:
* Single digit, starting with 1, or with 8
* Two-digit, starting with 1 or with 8
* < bracount
* > bracount
* read_number() overflow, starting with 1, or with 8

Everything... seems to work, I think? So this PR is equally correct as mine, it just adds some extra conditional logic.

Reply:
Was trying to minimize the change, but hopefully the new version (which does a little more refactoring to clean the currently messy code) is clearer.

Historically, resetting of `ptr` was meant to be only done in preparation for the fallback which is why I changed the comment, but could had used a little more verbosity to the change.